packages/superset-ui-chart/src/components/SuperChart.tsx

```diff
@@ -0,0 +1,218 @@
+      input => input.chartType,
+      input => input.overrideTransformProps,
+      (chartType, overrideTransformProps) => {
+        if (chartType) {
+          const Renderer = createLoadableRenderer({
+            loader: {
+              Chart: () => componentRegistry.getAsPromise(chartType),
+              transformProps: overrideTransformProps
+                ? () => Promise.resolve(overrideTransformProps)
+                : () => transformPropsRegistry.getAsPromise(chartType),
+            },
+            loading: (loadingProps: LoadingProps) => this.renderLoading(loadingProps, chartType),
+            render: this.renderChart,
+          });
+
+          // Trigger preloading.
+          Renderer.preload();
+
+          return Renderer;
+        }
+
+        return EMPTY;
+      },
+    );
+  }
+
+  processChartProps: (
+    input: {
+      chartProps: ChartProps;
+      preTransformProps?: TransformFunction;
+      transformProps?: TransformFunction;
+      postTransformProps?: TransformFunction;
+    },
+  ) => any;
+
+  createLoadableRenderer: (
+    input: {
+      chartType: string;
+      overrideTransformProps?: TransformFunction;
+    },
+  ) => LoadableRenderer<RenderProps, LoadedModules> | (() => null);
+
+  renderChart(loaded: LoadedModules, props: RenderProps) {
+    const Chart = getModule<typeof React.Component>(loaded.Chart);
+    const transformProps = getModule<TransformFunction>(loaded.transformProps);
+    const { chartProps, preTransformProps, postTransformProps } = props;
+
+    return (
+      <Chart
+        {...this.processChartProps({
+          /* eslint-disable sort-keys */
+          chartProps,
+          preTransformProps,
+          transformProps,
+          postTransformProps,
+          /* eslint-enable sort-keys */
+        })}
+      />
+    );
+  }
+
+  renderLoading(loadingProps: LoadingProps, chartType: string) {
+    const { error } = loadingProps;
+
+    if (error) {
+      return (
+        <div className="alert alert-warning" role="alert">
+          <strong>ERROR</strong>&nbsp;
+          <code>chartType=&quot;{chartType}&quot;</code> &mdash;
+          {error.toString()}
+        </div>
+      );
+    }
+
+    return null;
+  }
+
+  render() {
+    const {
+      id,
+      className,
+      preTransformProps,
+      postTransformProps,
+      chartProps = BLANK_CHART_PROPS,
+      onRenderSuccess,
+      onRenderFailure,
+    } = this.props;
+
+    // Create LoadableRenderer and start preloading
+    // the lazy-loaded Chart components
+    const Renderer = this.createLoadableRenderer(this.props);
+
+    // Do not render if chartProps is set to null.
+    // but the pre-loading has been started in this.createLoadableRenderer
+    // to prepare for rendering once chartProps becomes available.
+    if (chartProps === null) {
+      return null;
+    }
+
+    const containerProps: {
+      id?: string;
+      className?: string;
+    } = {};
+    if (id) {
```

Comment:
This might still create `<div id="" className="">` eventhough `id` and `className` are `undefined`. The `if` condition ensure it will be just `<div>`, which is slightly cleaner.